packages/plugin-data/src/api/data/soql/queryFields.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { ensureJsonArray, ensureJsonMap, ensureString, isJsonArray, toJsonMap } from '@salesforce/ts-types';
+import { Connection } from '@salesforce/core';
+import { Tooling } from '@salesforce/core/lib/connection';
+
+/**
+ * Class to represent a field when describing the fields that make up a query result
+ */
+export class Field {
+  public name: string;
+
+  public constructor(name: string) {
+    this.name = name;
+  }
+}
+/**
+ * Class to represent a subquery field when describing the fields that make up a query result
+ */
+export class SubqueryField extends Field {
+  public fields: Field[] = [];
+}
+
+/**
+ * Class to represent function fields, i.e. result of using aggregation function, like avg.
+ */
+export class FunctionField extends Field {
+  public alias: string | undefined;
+}
+
+/**
+ * Utility to fetch the columns involved in a soql query.
+ *
+ * Columns are then transformed into one of three types, Field, SubqueryField and FunctionField. List of
+ * fields is returned as the product.
+ *
+ * @param connection
+ * @param query
+ */
+export const retrieveColumns = async (connection: Connection | Tooling, query: string): Promise<Field[]> => {
```

Comment:
This seems like it's specific to command output right?  Do you expect people to want this outside of a command context?